assets/javascripts/discourse/initializers/add-search-results.js

```diff
@@ -15,54 +15,107 @@ import { Promise } from "rsvp";
 
 const CACHE_KEY = "discourse-encrypt-cache";
 
-function addCacheItem(session, type, item) {
+function getCache(session) {
   let cache = session.get(CACHE_KEY);
   if (!cache) {
     session.set(CACHE_KEY, (cache = {}));
   }
+  return cache;
+}
 
+function addObjectToCache(cache, type, object) {
   if (!cache[type]) {
     cache[type] = [];
-  } else if (item.id) {
-    cache[type] = cache[type].filter((i) => i.id !== item.id);
   }
-  cache[type].push(item);
+  cache[type][object.id] = object;
 }
 
-function getOrFetchCache(session) {
-  const cache = session.get(CACHE_KEY);
-  if (cache) {
-    return Promise.resolve(cache);
-  }
+function addPostToCache(cache, post) {
+  post.topic_title_headline = null;
+  addObjectToCache(cache, "posts", post);
+}
 
-  return ajax("/encrypt/posts")
-    .then((result) => {
-      const promises = [];
+function addTopicToCache(cache, topic) {
+  putTopicKey(topic.id, topic.topic_key);
+  putTopicTitle(topic.id, topic.encrypted_title);
+  return getTopicTitle(topic.id).then((title) => {
+    topic.title = title;
+    topic.fancy_title = `${iconHTML("user-secret")} ${title}`;
+    topic.excerpt = null;
+    addObjectToCache(cache, "topics", topic);
+    return topic;
+  });
+}
 
-      result.posts.forEach((post) => {
-        post.topic_title_headline = null;
-        addCacheItem(session, "posts", post);
-      });
+function loadCache(cache) {
+  return ajax("/encrypt/posts").then((result) => {
+    const promises = [];
+    result.posts.forEach((post) => addPostToCache(cache, post));
+    result.topics.forEach((topic) => {
```

Comment:
Extremely minor but you could set `let promises = result.topics.map()` rather than making a variable and pushing to it.